fw/sock_srv.c

```diff
@@ -277,6 +277,9 @@ tfw_sock_srv_connect_try(TfwSrvConn *srv_conn)
 		if (r != -ESHUTDOWN)
 			T_ERR("Unable to initiate a connect to server: %d\n",
 				r);
+		bh_lock_sock(sk);
+		ss_do_close(sk, 0);
+		bh_unlock_sock(sk);
 		sock_put(sk);
 		SS_CALL(connection_drop, sk);
```

Comment:
Why not `ss_close()`? It calls also `sock_put()` and `connection_drop`. It's error prone and breaks the abstraction layers w/o a clear reason

Reply:
Because here socket state is SS_CLOSE, but we should not call ss_do_close for socket in this state in all other places except here. We can add special flag for such socket, but I think that this solution is better

Reply:
I don't understand
> Because here socket state is SS_CLOSE, but we should not call ss_do_close for socket in this state in all other places except here.

Yes, we discuss this particular place in `tfw_sock_srv_connect_try()`. 

My objection that here are 5 lines of code
```C
		bh_lock_sock(sk);
		ss_do_close(sk, 0);
		bh_unlock_sock(sk);
		sock_put(sk);
		SS_CALL(connection_drop, sk);
```
dealing with low-level socket API - this is subject for sock.c, no sock_src.c, which should care about connection management and work with connection abstractions.

What would be wrong if we replace these 5 lines with `ss_close()`?

Reply:
look at the `ss_tx_action` case `SS_CLOSE`. we check that:
```
if (!((1 << sk->sk_state)
			      & (TCPF_ESTABLISHED | TCPF_SYN_SENT
				 | TCPF_CLOSE_WAIT)))
```
and if this if is true don't call `ss_do_close`. It is necessary to prevent multiple socket closing. But socket in `tfw_sock_srv_connect_try` has CLOSE state and this if is true. So if we call ss_close from `tfw_sock_srv_connect_try` instead of this five lines of code we never drop connection and close socket! We can pass special flag in ss_tx_action but I prefer this way. If you don't think so ok I will change it